infoschema/slow_log_cache.go

```diff
@@ -0,0 +1,337 @@
+	}
+	// Append after cache tuples.
+	for i := range afterCacheTuples {
+		rows = append(rows, afterCacheTuples[i].convertToDatumRow())
+	}
+	// Update cache.
+	go s.updateCache(s.cache.filePath, afterCacheTuples)
+	return rows, nil
+}
+
+// parseSlowLogFileAndUpdateCache gets data by parse the slow log file, and then update the cache.
+func (s *slowLogReader) parseSlowLogFileAndUpdateCache(tz *time.Location, filePath string, offset int64, filterFn func(t time.Time) bool, bypassFn func(t time.Time) bool) ([][]types.Datum, error) {
+	tuples, err := parseSlowLogDataFromFile(tz, filePath, offset, filterFn, bypassFn)
+	if err != nil {
+		return nil, err
+	}
+	go s.updateCache(filePath, tuples)
+	return convertSlowLogTuplesToDatums(tuples), nil
+}
+
+// parseSlowLogFileBeforeCachedAndCheckTruncate parse slow log file data which before cached start time
+// and check the log file has been truncated.
+func (s *slowLogReader) parseSlowLogFileBeforeCachedAndCheckTruncate(tz *time.Location) ([]*slowQueryTuple, bool, error) {
+	cacheStartTime := s.cache.getStartTime()
+	first := true
+	hasTruncate := false
+	beforeCacheTuples, err := parseSlowLogDataFromFile(tz, s.cache.filePath, 0, func(t time.Time) bool {
+		if first {
+			first = false
+			hasTruncate = t.After(cacheStartTime)
+		}
+		return t.Before(cacheStartTime)
+	}, nil)
+	return beforeCacheTuples, hasTruncate, err
+}
+
+// getSlowLogRowsFromCache gets slow log rows from cache.
+func (s *slowLogReader) getSlowLogRowsFromCache(tz *time.Location, rows [][]types.Datum) [][]types.Datum {
+	s.cache.buf.iterate(func(d interface{}) bool {
+		tuple := d.(*slowQueryTuple)
+		rows = append(rows, tuple.convertToDatumRow())
+		return false
+	})
+	return rows
+}
+
+// parseSlowLogFileAfterCached parse slow log file data which after cached end time.
+func (s *slowLogReader) parseSlowLogFileAfterCached(tz *time.Location, cacheEndTime time.Time, offset int64) ([]*slowQueryTuple, error) {
+	afterCacheTuples, err := parseSlowLogDataFromFile(tz, s.cache.filePath, offset, nil, func(t time.Time) bool {
+		return t.Before(cacheEndTime) || t.Equal(cacheEndTime)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return afterCacheTuples, err
+}
+
+func (s *slowLogReader) updateCache(filePath string, tuples []*slowQueryTuple) {
+	s.Lock()
+	defer s.Unlock()
+	if s.cache == nil {
+		s.cache = newSlowLogBuffer(config.GetGlobalConfig().Log.SlowQueryFile, slowQueryBufferSize)
+	}
+	if filePath != s.cache.filePath {
+		logutil.Logger(context.Background()).Info("slow query cache file not match", zap.String("cache file", s.cache.filePath), zap.String("data file", filePath))
+		return
+	}
+	if s.cache.buf.isEmpty() {
+		for i := range tuples {
+			s.cache.buf.write(tuples[i])
+		}
+		return
+	}
+	cacheEndTuple := s.cache.getEndTuple()
+	for i := range tuples {
+		if tuples[i].time.Before(cacheEndTuple.time) {
+			continue
+		}
+		if tuples[i].equal(cacheEndTuple) {
+			continue
+		}
+
+		s.cache.buf.write(tuples[i])
```

Comment:
Should we update `cacheEndTuple` after this statement?

Reply:
Good catch.
But normally, there won't have duplicate `tuple` in `tuples`, so I think it is no need to do this.